wrappers/python/nntile/layer/llama_attention.py

```diff
@@ -1517,3 +1517,156 @@ def to_torch_with_grads(self) -> LlamaAttention_torch:
                 to_numpy(self.in_proj_bias_v.grad).T.flatten()
             )
         return torch_layer
+
+    def get_layer_forward_flops(self):
+        total_forward_flops = 0
+        # Compute Q_transposed
+        # Q_transposed = einsum('ijkl,lmn->ijkmn', W_Q, X_Q)
+        # gemm (kv_group_size, n_head_kv, head_size, n_emb)
+        # by (n_emb, n_seq, n_batch)
+        # into (kv_group_size, n_head_kv, head_size, n_seq, n_batch)
+        w_q_shape = self.w_q.value.shape
+        x_q_shape = self.x_q.value.shape
+        qt_flops = 2 * np.prod(w_q_shape) * np.prod(x_q_shape[1:])
+        total_forward_flops += qt_flops
+        # Compute K_transposed
+        # K_transposed = einsum('jkl,lmn->jkmn', W_K, X_K)
+        # gemm (n_head_kv, head_size, n_emb) by (n_emb, n_seq, n_batch) into
+        # (n_head_kv, head_size, n_seq, n_batch)
+        w_k_shape = self.w_k.value.shape
+        x_k_shape = self.x_k.value.shape
+        kt_flops = 2 * np.prod(w_k_shape) * np.prod(x_k_shape[1:])
+        total_forward_flops += kt_flops
+        # Compute V_transposed
+        # V_transposed = einsum('jkl,lmn->jkmn', W_V, X_V)
+        # gemm (n_head_kv, head_size, n_emb) by (n_emb, n_seq, n_batch) into
+        # (n_head_kv, head_size, n_seq, n_batch)
+        w_v_shape = self.w_v.value.shape
+        x_v_shape = self.x_v.value.shape
+        vt_flops = 2 * np.prod(w_v_shape) * np.prod(x_v_shape[1:])
+        total_forward_flops += vt_flops
+        # Compute tensor for softmax
+        # A = 1.0/sqrt(head_size) * einsum('jklbi,jmlbi->kmlbi', K_rep, Q_rope)
+        # single batched gemm
+        # (head_size, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        # by (head_size, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        # into (n_seq, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        k_rep_shape = self.k_rep.value.shape
+        q_rope_shape = self.q_rope.value.shape
+        a_flops = 2 * np.prod(k_rep_shape[:2]) * np.prod(q_rope_shape[1:])
+        total_forward_flops += a_flops
+        # Apply value tensor
+        # B = einsum('jklbi,kmlbi->jmlbi', V_rep, A)
+        # batched gemm
+        # (head_size, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        # by (n_seq, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        # into (head_size, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        v_rep_shape = self.v_rep.value.shape
+        a_shape = self.a.value.shape
+        b_flops = 2 * np.prod(a_shape) * v_rep_shape[0]
```

Comment:
although `shape[0]` and `shape[1]` of `v_rep` are the same, technically we have to multiply by `v_rep_shape[1]`. If we switch this shape some day in future then flops will be bugged.

Reply:
``v_rep_shape[0] == head_size`` and ``v_rep_shape[1] == n_seq``. So why are they the same?

Reply:
ok, second time looking at this. I myself did a mistake. You version is correct. I was thinking of `2 * np.prod(v_rep_shape) * a_shape[1]`.